clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -278,17 +277,36 @@
         TreeSet<String> sortedCurrentSubscriptions = new TreeSet<>(new SubscriptionComparator(currentAssignment));
         sortedCurrentSubscriptions.addAll(currentAssignment.keySet());
 
-        balance(currentAssignment, sortedPartitions, unassignedPartitions, sortedCurrentSubscriptions,
+        balance(sortedPartitions, unassignedPartitions, sortedCurrentSubscriptions,
                 consumer2AllPotentialPartitions, partition2AllPotentialConsumers, currentPartitionConsumer);
         return currentAssignment;
     }
 
-    private void prepopulateCurrentAssignments(Map<String, Subscription> subscriptions,
-                                               Map<String, List<TopicPartition>> currentAssignment) {
+    private void prepopulateCurrentAssignments(Map<String, Subscription> subscriptions) {
+        Map<TopicPartition, String> currentPartitionToMemberAssignment = new HashMap<>();
+        currentAssignment.forEach((member, partitions) -> partitions.stream().forEach(
+            partition -> currentPartitionToMemberAssignment.put(partition, member)
+        ));
+
+        currentAssignment.clear();
+        mainLoop:
         for (Map.Entry<String, Subscription> subscriptionEntry : subscriptions.entrySet()) {
             ByteBuffer userData = subscriptionEntry.getValue().userData();
-            if (userData != null && userData.hasRemaining())
-                currentAssignment.put(subscriptionEntry.getKey(), deserializeTopicPartitionAssignment(userData));
+            if (userData != null && userData.hasRemaining()) {
+                String member = subscriptionEntry.getKey();
+                List<TopicPartition> deserializedAssignment = deserializeTopicPartitionAssignment(userData);
+                for (TopicPartition partition: deserializedAssignment) {
+                    // this check is the reason for having the leader cache the current assignment
+                    if (currentPartitionToMemberAssignment.containsKey(partition) && !currentPartitionToMemberAssignment.get(partition).equals(member)) {
+                        log.debug("Member " + member + " seems to be out-of-date with respect to its previous assignment " +
+                                "(maybe it's rejoining the group after some pause?). Therefore, It's previous assignment " +
+                                "will be ignored.");
+                        continue mainLoop;
+                    }
+                }
+
+                currentAssignment.put(member, deserializedAssignment);
```

Comment:
This caching logic might work if the leader does not change.

But I have the impression that it is possible to end up in the same problem when the leader changes.  Because as far as I see, the new leader will not have build up its cache.  (or even worse has an outdated cache from a previous time it was leader)

Which I think can still cause troubles.

I see two other possible solutions:
- The easiest one to implement I think is keep the previous code, but change the type of the `currentAssignment` parameter from a `Map<String, List<TopicPartition>>` to a `Map<TopicPartition, String>`, or keep the type but do a post-processing step on this list where double TopicPartitions are removed before doing the `sortPartitions`.

- Another possibility is putting a generation counter in the schema, and only keep assignments with the maximum generation counter.  (or those where it is still missing for backwards compatibility).

The first possible solutions is easy to implement but makes it a little less sticky. (which I do not mind)
The second solution is more sound, but implementing it in a backwards compatible way is a challenge.